File: buerguer-pig/HamburgesasEIngredientes-Java-IvanRC/src/main/java/repository/ingrediente/IngredienteRepositoryImpl.java
package repository.ingrediente;

import models.Ingrediente;
import storage.ingrediente.IngredienteStorageService;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class IngredienteRepositoryImpl implements IngredienteRepository {

    private final IngredienteStorageService storage;

    public IngredienteRepositoryImpl(IngredienteStorageService storage) {
        this.storage = storage;
    }

    private ArrayList<Ingrediente> ingredientes = new ArrayList<>();

    @Override
    public List<Ingrediente> getAll() throws Exception {
        loadAll();
        return ingredientes;
    }

    @Override
    public Ingrediente getById(Integer id) throws Exception {
        loadAll();
        return ingredientes.stream().filter(ingrediente -> ingrediente.getId() == id).findFirst().orElse(null);
    }

    @Override
    public Ingrediente add(Ingrediente entity) throws Exception {
        Ingrediente ingrediente = getById(entity.getId());
        if(ingrediente == null){
            ingredientes.add(entity);
            saveAll();
            return entity;
        }
        return null;
    }

    @Override
    public Ingrediente update(Ingrediente entity) {
        return null;
    }

    @Override
    public Ingrediente remove(Integer id) {
        return null;
    }

    @Override
    public List<Ingrediente> getByName(String name) {
        return null;
    }


    private void loadAll() throws Exception {
        ingredientes.clear();
        storage.loadAll().forEach(ingrediente ->
            ingredientes.add(ingrediente)
        );
    }

    private void saveAll() throws Exception {
        storage.saveAll(ingredientes);
    }
}

File: buerguer-pig/HamburgesasEIngredientes-Kotlin-IvanRC/src/main/kotlin/repository/Ingrediente/IngredienteRepositoryImpl.kt
package repository.Ingrediente

import model.Ingrediente
import mu.KotlinLogging
import storageService.base.StorageService

class IngredienteRepositoryImpl(
    private val storage: StorageService<Ingrediente>
): IngredienteRepository {

    private val logger = KotlinLogging.logger {}
    private var ingredientes = mutableListOf<Ingrediente>()

    override fun findByName(name: String): List<Ingrediente> {
        logger.debug { "Repository: Se buscan ingredientes según su nombre." }
        TODO("Not yet implemented")
    }

    override fun findById(id: Int): Ingrediente? {
        logger.debug { "Repository: Se busca un ingrediente según su id." }
        loadAll()
        return ingredientes.find { it.id == id }
    }

    override fun getAll(): List<Ingrediente> {
        logger.debug { "Repository: Se consiguen ingredientes." }
        loadAll()
        return ingredientes.toList()
    }

    override fun add(entity: Ingrediente): Ingrediente? {
        logger.debug { "Repository: Se añade un ingrediente." }
        (findById(entity.id))?:run{
            ingredientes.add(entity)
            saveAll()
            return entity
        }
        return null
    }

    override fun delete(id: Int): Ingrediente? {
        logger.debug { "Repository: Se elimina un ingrediente." }
        TODO("Not yet implemented")
    }

    override fun update(entity: Ingrediente): Ingrediente? {
        logger.debug { "Repository: Se actualiza un ingrediente." }
        TODO("Not yet implemented")
    }

    fun saveAll(){
        logger.debug { "Repository: Se guargan los ingredientes en el storage." }
        storage.saveAll(ingredientes)
    }

    fun loadAll(){
        logger.debug { "Repository: Se cargan los ingredientes en el storage." }
        ingredientes = storage.loadAll().toMutableList()
    }
}